File: java/src/main/java/io/cucumber/gherkin/TokenMatcher.java
package io.cucumber.gherkin;

import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import io.cucumber.messages.types.StepKeywordType;

import static io.cucumber.gherkin.GherkinLanguageConstants.DOCSTRING_ALTERNATIVE_SEPARATOR;
import static io.cucumber.gherkin.GherkinLanguageConstants.DOCSTRING_SEPARATOR;
import static io.cucumber.gherkin.Parser.ITokenMatcher;
import static io.cucumber.gherkin.Parser.TokenType;

class TokenMatcher implements ITokenMatcher {
    private static final Pattern LANGUAGE_PATTERN = Pattern.compile("^\\s*#\\s*language\\s*:\\s*([a-zA-Z\\-_]+)\\s*$");
    private final GherkinDialectProvider dialectProvider;
    private GherkinDialect currentDialect;
    private String activeDocStringSeparator = null;
    private int indentToRemove = 0;

    public TokenMatcher(GherkinDialectProvider dialectProvider) {
        this.dialectProvider = dialectProvider;
        reset();
    }

    public TokenMatcher() {
        this(new GherkinDialectProvider());
    }

    public TokenMatcher(String defaultDialectName) {
        this(new GherkinDialectProvider(defaultDialectName));
    }

    @Override
    public void reset() {
        activeDocStringSeparator = null;
        indentToRemove = 0;
        currentDialect = dialectProvider.getDefaultDialect();
    }

    private GherkinDialect getCurrentDialect() {
        return currentDialect;
    }

    private void setTokenMatched(Token token, TokenType matchedType, String text, String keyword, Integer indent, StepKeywordType keywordType, List<GherkinLineSpan> items) {
        token.matchedType = matchedType;
        token.matchedKeyword = keyword;
        token.keywordType = keywordType;
        token.matchedText = text;
        token.matchedItems = items;
        token.matchedGherkinDialect = getCurrentDialect();
        token.matchedIndent = indent != null ? indent : (token.line == null ? 0 : token.line.indent());
        token.location = new Location(token.location.getLine(), token.matchedIndent + 1);
    }

    @Override
    public boolean match_EOF(Token token) {
        if (token.isEOF()) {
            setTokenMatched(token, TokenType.EOF, null, null, null, null, null);
            return true;
        }
        return false;
    }

    @Override
    public boolean match_Other(Token token) {
        String text = token.line.getLineText(indentToRemove); //take the entire line, except removing DocString indents
        setTokenMatched(token, TokenType.Other, unescapeDocString(text), null, 0, null, null);
        return true;
    }

    @Override
    public boolean match_Empty(Token token) {
        if (token.line.isEmpty()) {
            setTokenMatched(token, TokenType.Empty, null, null, null, null, null);
            return true;
        }
        return false;
    }

    @Override
    public boolean match_Comment(Token token) {
        if (token.line.startsWith(GherkinLanguageConstants.COMMENT_PREFIX)) {
            String text = token.line.getLineText(0); //take the entire line
            setTokenMatched(token, TokenType.Comment, text, null, 0, null, null);
            return true;
        }
        return false;
    }

    @Override
    public boolean match_Language(Token token) {
        Matcher matcher = LANGUAGE_PATTERN.matcher(token.line.getLineText(0));
        if (matcher.matches()) {
            String language = matcher.group(1);
            setTokenMatched(token, TokenType.Language, language, null, null, null, null);

            currentDialect = dialectProvider.getDialect(language)
                    .orElseThrow(() -> new ParserException.NoSuchLanguageException(language, token.location));
            return true;
        }
        return false;
    }

    @Override
    public boolean match_TagLine(Token token) {
        if (token.line.startsWith(GherkinLanguageConstants.TAG_PREFIX)) {
            setTokenMatched(token, TokenType.TagLine, null, null, null, null, token.line.getTags());
            return true;
        }
        return false;
    }

    @Override
    public boolean match_FeatureLine(Token token) {
        return matchTitleLine(token, TokenType.FeatureLine, currentDialect.getFeatureKeywords());
    }

    @Override
    public boolean match_RuleLine(Token token) {
        return matchTitleLine(token, TokenType.RuleLine, currentDialect.getRuleKeywords());
    }

    @Override
    public boolean match_BackgroundLine(Token token) {
        return matchTitleLine(token, TokenType.BackgroundLine, currentDialect.getBackgroundKeywords());
    }

    @Override
    public boolean match_ScenarioLine(Token token) {
        return matchTitleLine(token, TokenType.ScenarioLine, currentDialect.getScenarioKeywords()) ||
                matchTitleLine(token, TokenType.ScenarioLine, currentDialect.getScenarioOutlineKeywords());
    }

    @Override
    public boolean match_ExamplesLine(Token token) {
        return matchTitleLine(token, TokenType.ExamplesLine, currentDialect.getExamplesKeywords());
    }

    private boolean matchTitleLine(Token token, TokenType tokenType, List<String> keywords) {
        for (String keyword : keywords) {
            if (token.line.startsWithTitleKeyword(keyword)) {
                String title = token.line.getRestTrimmed(keyword.length() + GherkinLanguageConstants.TITLE_KEYWORD_SEPARATOR.length());
                setTokenMatched(token, tokenType, title, keyword, null, null, null);
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean match_DocStringSeparator(Token token) {
        return activeDocStringSeparator == null
                // open
                ? match_DocStringSeparator(token, DOCSTRING_SEPARATOR, true) ||
                match_DocStringSeparator(token, GherkinLanguageConstants.DOCSTRING_ALTERNATIVE_SEPARATOR, true)
                // close
                : match_DocStringSeparator(token, activeDocStringSeparator, false);
    }

    private boolean match_DocStringSeparator(Token token, String separator, boolean isOpen) {
        if (token.line.startsWith(separator)) {
            String mediaType = null;
            if (isOpen) {
                mediaType = token.line.getRestTrimmed(separator.length());
                activeDocStringSeparator = separator;
                indentToRemove = token.line.indent();
            } else {
                activeDocStringSeparator = null;
                indentToRemove = 0;
            }

            setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, null, null, null);
            return true;
        }
        return false;
    }

    @Override
    public boolean match_StepLine(Token token) {
        List<String> keywords = currentDialect.getStepKeywords();
        for (String keyword : keywords) {
            if (token.line.startsWith(keyword)) {
                String stepText = token.line.getRestTrimmed(keyword.length());
                StepKeywordType keywordType = getKeywordType(keyword);
                setTokenMatched(token, TokenType.StepLine, stepText, keyword, null, keywordType, null);
                return true;
            }
        }
        return false;
    }

    private StepKeywordType getKeywordType(String stepKeyword) {
        Set<StepKeywordType> keywordTypes = currentDialect.getStepKeywordTypesSet(stepKeyword);
        if (keywordTypes.size() == 1) {
            return keywordTypes.iterator().next();
        }
        return StepKeywordType.UNKNOWN;
    }

    @Override
    public boolean match_TableRow(Token token) {
        if (token.line.startsWith(GherkinLanguageConstants.TABLE_CELL_SEPARATOR)) {
            setTokenMatched(token, TokenType.TableRow, null, null, null, null, token.line.getTableCells());
            return true;
        }
        return false;
    }

    private String unescapeDocString(String text) {
        if (DOCSTRING_SEPARATOR.equals(activeDocStringSeparator)) {
            return text.replace("\\\"\\\"\\\"", DOCSTRING_SEPARATOR);
        }

        if (DOCSTRING_ALTERNATIVE_SEPARATOR.equals(activeDocStringSeparator)) {
            return text.replace("\\`\\`\\`", DOCSTRING_ALTERNATIVE_SEPARATOR);
        }
        return text;
    }
}

File: dart/lib/src/language/token_matcher.dart
import 'package:gherkin/parser.dart';
import 'package:cucumber_messages/cucumber_messages.dart' as messages;

import 'gherkin_language_constants.dart';
import 'gherkin_line_span.dart';
import 'i_gherkin_dialect.dart';
import 'i_gherkin_dialect_provider.dart';
import 'i_gherkin_line.dart';
import 'location.dart';
import 'token.dart';
import 'token_type.dart';
import 'package:gherkin/extensions.dart';

class TokenMatcher implements ITokenMatcher {
  static final RegExp _languagePattern = RegExp(
    r'^\s*#\s*language\s*:\s*([a-zA-Z\-_]+)\s*$',
  );

  final IGherkinDialectProvider _dialectProvider;

  IGherkinDialect _currentDialect;
  String _activeDocStringSeparator = Strings.empty;
  int _indentToRemove = 0;
  Map<String, List<messages.StepKeywordType>> _keywordTypesMap = {};

  TokenMatcher(this._dialectProvider)
    : _currentDialect = _dialectProvider.defaultDialect {
    _initializeKeywordTypes();
  }

  IGherkinDialect get currentDialect {
    if (_currentDialect.isEmpty) {
      return _dialectProvider.defaultDialect;
    }
    return _currentDialect;
  }

  @override
  void reset() {
    _activeDocStringSeparator = Strings.empty;
    _indentToRemove = 0;
    _currentDialect = _dialectProvider.defaultDialect;
    _initializeKeywordTypes();
  }

  @override
  bool matchEof(Token token) {
    if (token.isEof) {
      setTokenMatched(token, TokenType.eof);
      return true;
    }
    return false;
  }

  @override
  bool matchOther(Token token) {
    /// take the entire line, except removing DocString indents
    var text = token.line.getLineText(_indentToRemove);
    setTokenMatched(
      token,
      TokenType.other,
      text: _unescapeDocString(text),
      indent: 0,
    );
    return true;
  }

  @override
  bool matchEmpty(Token token) {
    if (token.line.isEmptyLine) {
      setTokenMatched(token, TokenType.empty);
      return true;
    }
    return false;
  }

  @override
  bool matchComment(Token token) {
    if (token.line.startsWith(GherkinLanguageConstants.commentPrefix)) {
      var text = token.line.getLineText(IGherkinLine.entireLine);
      setTokenMatched(token, TokenType.comment, text: text, indent: 0);
      return true;
    }
    return false;
  }

  @override
  bool matchLanguage(Token token) {
    var match = _languagePattern.firstMatch(
      token.line.getLineText(IGherkinLine.entireLine),
    );
    if (match != null) {
      var language = match.group(1) ?? Strings.empty;
      setTokenMatched(token, TokenType.language, text: language);
      _currentDialect = _dialectProvider.getDialect(language, token.location);
      _initializeKeywordTypes();
      return true;
    }
    return false;
  }

  @override
  bool matchTagLine(Token token) {
    if (token.line.startsWith(GherkinLanguageConstants.tagPrefix)) {
      setTokenMatched(token, TokenType.tagLine, items: token.line.tags);
      return true;
    }
    return false;
  }

  @override
  bool matchFeatureLine(Token token) => _matchTitleLine(
    token,
    TokenType.featureLine,
    currentDialect.featureKeywords,
  );

  @override
  bool matchRuleLine(Token token) =>
      _matchTitleLine(token, TokenType.ruleLine, currentDialect.ruleKeywords);

  @override
  bool matchBackgroundLine(Token token) => _matchTitleLine(
    token,
    TokenType.backgroundLine,
    currentDialect.backgroundKeywords,
  );

  @override
  bool matchScenarioLine(Token token) =>
      _matchTitleLine(
        token,
        TokenType.scenarioLine,
        currentDialect.scenarioKeywords,
      ) ||
      _matchTitleLine(
        token,
        TokenType.scenarioLine,
        currentDialect.scenarioOutlineKeywords,
      );

  @override
  bool matchExamplesLine(Token token) => _matchTitleLine(
    token,
    TokenType.examplesLine,
    currentDialect.examplesKeywords,
  );

  bool _matchTitleLine(
    Token token,
    TokenType tokenType,
    List<String> keywords,
  ) {
    for (var keyword in keywords) {
      if (token.line.startsWithTitleKeyword(keyword)) {
        var title = token.line.getRestTrimmed(
          keyword.length +
              GherkinLanguageConstants.titleKeywordSeparator.length,
        );
        setTokenMatched(token, tokenType, keyword: keyword, text: title);
        return true;
      }
    }
    return false;
  }

  @override
  bool matchDocStringSeparator(Token token) =>
      _activeDocStringSeparator.isEmpty
          // open
          ? _matchDocStringSeparator(
                token,
                GherkinLanguageConstants.docStringSeparator,
                true,
              ) ||
              _matchDocStringSeparator(
                token,
                GherkinLanguageConstants.docStringAlternativeSeparator,
                true,
              )
          // close
          : _matchDocStringSeparator(token, _activeDocStringSeparator, false);

  bool _matchDocStringSeparator(Token token, String separator, bool isOpen) {
    if (token.line.startsWith(separator)) {
      var mediaType = Strings.empty;
      if (isOpen) {
        mediaType = token.line.getRestTrimmed(separator.length);
        _activeDocStringSeparator = separator;
        _indentToRemove = token.line.indent;
      } else {
        _activeDocStringSeparator = Strings.empty;
        _indentToRemove = 0;
      }
      setTokenMatched(
        token,
        TokenType.docStringSeparator,
        text: mediaType,
        keyword: separator,
      );
      return true;
    }
    return false;
  }

  @override
  bool matchStepLine(token) {
    var keywords =
        currentDialect.stepKeywords.toList()
          ..sort((a, b) => b.length - a.length);
    for (var keyword in keywords) {
      if (token.line.startsWith(keyword)) {
        var stepText = token.line.getRestTrimmed(keyword.length);
        setTokenMatched(
          token,
          TokenType.stepLine,
          keyword: keyword,
          text: stepText,
          keywordType: _keywordType(keyword),
        );
        return true;
      }
    }
    return false;
  }

  @override
  bool matchTableRow(token) {
    if (token.line.startsWith(GherkinLanguageConstants.tableCellSeparator)) {
      setTokenMatched(token, TokenType.tableRow, items: token.line.tableCells);
      return true;
    }
    return false;
  }

  void setTokenMatched(
    Token token,
    TokenType matchedType, {
    String text = Strings.empty,
    String keyword = Strings.empty,
    messages.StepKeywordType? keywordType,
    int indent = Int.min,
    Iterable<GherkinLineSpan> items = const <GherkinLineSpan>[],
  }) {
    token.matchedType = matchedType;
    token.matchedKeyword = keyword;
    token.matchedKeywordType = keywordType;
    token.matchedText = text;
    token.matchedItems = items;
    token.matchedGherkinDialect = currentDialect;
    token.matchedIndent =
        matchedType == TokenType.empty
            ? 0
            : indent.isNotMin
            ? indent
            : (token.line.isEmpty ? 0 : token.line.indent);
    token.location = Location(token.location.line, token.matchedIndent + 1);
  }

  void _initializeKeywordTypes() {
    _keywordTypesMap = {};
    _addKeywordTypeMappings(
      currentDialect.givenStepKeywords,
      messages.StepKeywordType.context,
    );
    _addKeywordTypeMappings(
      currentDialect.whenStepKeywords,
      messages.StepKeywordType.action,
    );
    _addKeywordTypeMappings(
      currentDialect.thenStepKeywords,
      messages.StepKeywordType.outcome,
    );
    _addKeywordTypeMappings(
      currentDialect.andStepKeywords,
      messages.StepKeywordType.conjunction,
    );
    _addKeywordTypeMappings(
      currentDialect.butStepKeywords,
      messages.StepKeywordType.conjunction,
    );
  }

  void _addKeywordTypeMappings(
    Iterable<String> keywords,
    messages.StepKeywordType keywordType,
  ) {
    for (final keyword in keywords) {
      _keywordTypesMap
          .putIfAbsent(keyword, () => <messages.StepKeywordType>[])
          .add(keywordType);
    }
  }

  messages.StepKeywordType _keywordType(String keyword) {
    final keywordTypes = _keywordTypesMap[keyword];
    if (keywordTypes == null || keywordTypes.length != 1) {
      return messages.StepKeywordType.unknown;
    }
    return keywordTypes.single;
  }

  String _unescapeDocString(String text) {
    if (GherkinLanguageConstants.docStringSeparator ==
        _activeDocStringSeparator) {
      return text.replaceFirst(
        '\\"\\"\\"',
        GherkinLanguageConstants.docStringSeparator,
      );
    }
    if (GherkinLanguageConstants.docStringAlternativeSeparator ==
        _activeDocStringSeparator) {
      return text.replaceFirst(
        '\\`\\`\\`',
        GherkinLanguageConstants.docStringAlternativeSeparator,
      );
    }
    return text;
  }
}
